### PR TITLE
Remove PyCrypto

### LINKED ISSUE #72646: allow to assign a dictionary to itself when used with combine filter
##### SUMMARY

combine filter cannot be assigned to the same hash key that is applied, basically on itself,  to just subsite a part of a dictionary. Would be nice to have this feature, because at the moment is achievable only changing to `hash_behavior=merge` that is deprecated and "unwanted"

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

combine filter to give the possiblity to do something like

##### ADDITIONAL INFORMATION

the example above or similar, when using the destination key as an input, independently from combine filter, go on recursive loop error, when should not be the case. 

(It could be a bug)

```yaml
---
defaultRoleAVarsDict:
  a: "a"
  b: "c"
---
roles:
  role: roleA
  vars:
    defaultRoleAVarsDict: "{{ defaultRoleAVarsDict | combine({'b':'be'}) }}"
```

### LINKED ISSUE #74626: Drop `PyCrypto` support.
##### SUMMARY

`PyCrypto` is unmaintained, with the last commit to its repository in 2014. It no longer works as of Python 3.10 due to changes in arg parsing: https://docs.python.org/3.10/c-api/arg.html#arg-parsing

What little code remains in Ansible to support `PyCrypto` should be removed.

##### ISSUE TYPE

Feature Idea

##### COMPONENT NAME

ansible

### PR DESCRIPTION
**Pull request by samdoran**
##### SUMMARY

Remove `PyCrypto` from Ansible since it is no longer maintained and not compatible with Python 3.10.

Fixes #74626.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

`packaging/debian/rules`
`setup.py`
`test/units/parsing/vault`
`lib/ansible/executor/process/worker.py`
`lib/ansible/parsing/vault/__init__.py`
`docs/docsite/rst/installation_guide/intro_installation.rst`

### CODE REVIEW

**Review by mattclay — APPROVED** [review]

**Review by nitzmahone — APPROVED** [review]

File: lib/ansible/parsing/vault/__init__.py [hunk 1]
```diff
@@ -1310,11 +1237,9 @@ def _decrypt_cryptography(cls, b_ciphertext, b_crypted_hmac, b_key1, b_key2, b_i
     @staticmethod
     def _is_equal(b_a, b_b):
         """
-        Comparing 2 byte arrrays in constant time
-        to avoid timing attacks.
+        Comparing 2 byte arrays in constant time to avoid timing attacks.
 
-        It would be nice if there was a library for this but
-        hey.
+        It would be nice if there was a library for this but hey.
```
Comment (relrod) [inline lib/ansible/parsing/vault/__init__.py]:
```suggestion
        It would be nice if there were a library for this but hey.
```

**Review by relrod — APPROVED** [review]

**Comment by acozine** [comment]
Currently we list `cryptography` as an optional add-on for Vault, a way to speed it up: https://docs.ansible.com/ansible/latest/user_guide/vault.html#speeding-up-ansible-vault

Is it now a pre-requisite for using Vault? If so, we should update that page.

**Comment by nitzmahone** [comment]
`cryptography` has been a required dependency for Ansible itself for awhile now, so we can probably just get rid of any verbiage about it at all in the vault docs...

**Comment by samdoran** [comment]
I removed the section in the Ansible Vault docs about requiring `cryptography`.

**Comment by felixfontein** [comment]
@samdoran the issue mentioned in the changelog fragment is #72646 (which is unrelated to PyCrypto), not #74626 as mentioned in this PR.

File: docs/docsite/rst/porting_guides/porting_guide_core_2.12.rst [hunk 2]
```diff
@@ -25,7 +25,7 @@ No notable changes
 Command Line
 ============
 
-No notable changes
+* ``ansible-vault`` no longer supports ``PyCrypto`` and requires ``cryptograhpy``.
```
Comment (felixfontein) [inline docs/docsite/rst/porting_guides/porting_guide_core_2.12.rst:28]: Typo: `cryptography`

Comment (samdoran) [reply docs/docsite/rst/porting_guides/porting_guide_core_2.12.rst:28]: 🤦‍♂️

**Pull request merged**